### ISSUE TITLE
Endless loop in syntax check with 5.8.5

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7509]
Migrated from [rt.perl.org#31691](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31691) (status was 'resolved')

Searchable as RT31691$

**Comment by p5pRT** [comment]
### From aspa@merlot.kronodoc.fi

#### Created by aspa@merlot.kronodoc.fi



trying to run the following code fragment through the syntax checker results
in something that seems like an endless loop\. Perl 5\.8\.3 is able to do the
syntax check in the code\.

my \($fields\, $edited\);
foreach my $fid \(keys %$fields\) \{
&nbsp;    if\(grep $fid \($edited\)\) \{
&nbsp;    print "foo";
&nbsp;    \}
\}

the syntax for grep is invalid but the result is still not what i'd expect\.

best regards\,

&nbsp;    aspa

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.5:

Configured by aspa at Thu Sep 23 12:59:47 EEST 2004.

Summary of my perl5 (revision 5 version 8 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.4.21-15.0.4.elsmp, archname=i686-linux
    uname='linux merlot.kronodoc.fi 2.4.21-15.0.4.elsmp #1 smp sat jul 31 01:25:25 edt 2004 i686 i686 i386 gnulinux '
    config_args='-d -e -Dprefix=/opt/kronodoc/perl/kb21 -Dcc=gcc -Duseshrplib -Uinstallusrbinperl -Dusemymalloc -Accflags=-DUSE_HASH_SEED_EXPLICIT -mcpu=i686'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DUSE_HASH_SEED_EXPLICIT -mcpu=i686 -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2',
    cppflags='-DUSE_HASH_SEED_EXPLICIT -mcpu=i686 -fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.3 20030502 (Red Hat Linux 3.2.3-42)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/opt/kronodoc/perl/kb21/lib/5.8.5/i686-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.5:
    /opt/kronodoc/perl/kb21/lib/5.8.5/i686-linux
    /opt/kronodoc/perl/kb21/lib/5.8.5
    /opt/kronodoc/perl/kb21/lib/site_perl/5.8.5/i686-linux
    /opt/kronodoc/perl/kb21/lib/site_perl/5.8.5
    /opt/kronodoc/perl/kb21/lib/site_perl
    .


Environment for perl v5.8.5:
    HOME=/home/aspa
    LANG=en_US.iso885915
    LANGUAGE (unset)
    LANGVAR=en_US.iso885915
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

gdb was misbehaving for me\, but it appears that the endless loop is
occuring in the function Perl\_ck\_grep\(\)\.  Now\, its time to recompile
with a \-g\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Sep 23\, 2004 at 04&#8203;:05&#8203;:22PM \-0000\, Steve Peters via RT wrote&#8203;:
> gdb was misbehaving for me\, but it appears that the endless loop is
> occuring in the function Perl\_ck\_grep\(\)\.  Now\, its time to recompile
> with a \-g\.

It's the op\_next field in the padsv op below pointing to itself\.

At some points during compilation it's normal practice to make an op point
to itself; this is usually undone later\. I've never understood what it's
supposed to signify\.

Dave\.

\-\- 
The perl5 internals are a complete mess\. It's like Jenga \- to get the
perl5 tower taller and do something new you select a block somewhere in
the middle\, with trepidation pull it out slowly\, and then carefully
balance it somewhere new\, hoping the whole edifice won't collapse as a
result\.
&nbsp;        \- Nicholas Clark\, based on an original by Simon Cozens\.

**Comment by p5pRT** [comment]
### From @iabyn

fixed by f6435df36c4167f97876032948ee21a446b75f27

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
